### ISSUE TITLE
Finish compact decimal formatting

### ISSUE THREAD

**Issue by sffc** [unicode-org/icu4x#3647]
We have formatting from integers. We need to add it from general FixedDecmial & f64 and add support for currencies.

**Comment by sffc** [comment]
Note that there are two places where floating point numbers are rounded: first in the conversion of the floating point number to the shortest decimal string, and second when performing the user-visible rounding logic.

Discussion about double rounding:

- @sffc - Double-rounding is specified [here](https://tc39.es/ecma402/#sec-tointlmathematicalvalue) in EMCA-402. In ECMAScript, a way to rationalize this is that floating point numbers are seen as compact representations of decimal numbers, so this operation is simply a function in the 1:1 mapping.
- @eggrobin - In the general case, that rationalization is not applicable. But, in the case of number formatting in ICU4X, this is a definite edge case. We could say that we can change the behavior here in the future if desired. I don't want to tie ourselves to the ECMA double rounding standard.

Consensus: For now, implement compact rounding with the ECMA-style double rounding, but leave open the door to do it differently in the future.

LGTM: @eggrobin @sffc @skius @younies

**Comment by younies** [comment]
is this fixed ?

**Comment by sffc** [comment]
No, @robertbastian was working on this but it needs a graduation review.

**Comment by Manishearth** [comment]
The current status of this code is that it is good enough for a tech preview. We should graduate next release but do not need to for 2.2

**Comment by sffc** [comment]
For 2.3, I'd like to see the new CompactDecimalFormatter. It would also be nice to have the new DecimalFormatter. Both should be unstable for 2.3

If you make the new DecimalFormatter, also please do Nan/Infinity the way you think they should be

**Comment by sffc** [comment]
Moving this to 2.4
